### PR TITLE
PS-7764: Implement ZenFS for binary tarballs and create a separate package for Debian

### PR DESCRIPTION
**Pull request by Sudokamikaze**
Builds: https://rel.cd.percona.com/job/ps8.0-autobuild-illiatests/3/

### CODE REVIEW

**Comment by george-lorch** [comment]
Excellent to see this moving along, thanks!

**Comment by Sudokamikaze** [comment]
> Excellent to see this moving along, thanks!

Tarballs are ready on my AWS instance, we'll discuss with Evgeniy approach for Debian packages on tomorrow's DS

File: build-ps/build-binary.sh [hunk 1]
```diff
@@ -423,28 +432,63 @@ fi
...
+            cp $SOURCEDIR/plugin/zenfs/util/zenfs $INSTALLDIR/
+            rm -rf $INSTALL_ROOT $BUILD_ROOT $SOURCEDIR
+        fi
+    }
+
     cd "$INSTALLDIR/usr/local/"
     #PS-4854 Percona Server for MySQL tarball without AGPLv3 dependency/license
     find $PRODUCT_FULL -type f -name 'COPYING.AGPLv3' -delete
+    if [[ ${WITH_ZENFS} == "ON" ]]; then
+        build_and_place_zenfs_utils
```
Comment (EvgeniyPatlan) [inline build-ps/build-binary.sh]: shouldn't we just call this function once and then only copy zenfs bin file here and on line 490?
It should be called on the build stage not on packaging stage

File: build-ps/debian/rules_zenfs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+		-DROCKSDB_DISABLE_AVX2=1 \
+		-DROCKSDB_DISABLE_MARCH_NATIVE=1 \
+		-DFORCE_INSOURCE_BUILD=1 \
+		-DDOWNLOAD_BOOST=1 \
+		-DWITH_BOOST=libboost \
+		-DWITH_SYSTEM_LIBS=ON \
+                -DWITH_PROTOBUF=bundled \
+                -DWITH_RAPIDJSON=bundled \
+                -DWITH_ICU=bundled \
+                -DWITH_ZSTD=bundled \
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules_zenfs]: indentation

File: build-ps/debian/rules_zenfs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+		-DWITH_ZLIB=bundled \
+                -DWITH_LZ4=bundled \
+                -DWITH_EDITLINE=bundled \
+                -DWITH_LIBEVENT=bundled \
+                -DWITH_NUMA=ON \
+                -DWITH_LDAP=system \
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules_zenfs]: indentation

File: build-ps/debian/rules_zenfs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		-DROCKSDB_DISABLE_AVX2=1 \
+		-DROCKSDB_DISABLE_MARCH_NATIVE=1 \
+		-DDOWNLOAD_BOOST=1 \
+		-DWITH_BOOST=libboost \
+		-DWITH_SYSTEM_LIBS=ON \
+                -DWITH_PROTOBUF=bundled \
+                -DWITH_RAPIDJSON=bundled \
+                -DWITH_ICU=bundled \
+                -DWITH_ZSTD=bundled \
+                -DWITH_LZ4=bundled \
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules_zenfs]: indebtation

File: build-ps/debian/rules_zenfs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		-DWITH_ZLIB=bundled \
+                -DWITH_EDITLINE=bundled \
+                -DWITH_LIBEVENT=bundled \
+                -DWITH_NUMA=ON \
+                -DWITH_LDAP=system \
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules_zenfs]: indentation

File: build-ps/debian/rules_zenfs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+override_dh_strip:
+	dh_strip --dbg-package=percona-server-dbg
+	dh_strip -Xlibprotobuf-lite
+
+override_dh_installinit:
+	@echo "RULES.$@"
+#	dh_apparmor -ppercona-server-server --profile-name=usr.sbin.mysqld
+#	dh_apparmor -ppercona-mysql-router --profile-name=usr.bin.mysqlrouter
+	if [ "$(DISTRELEASE)" != "precise" -a "$(DISTRELEASE)" != "trusty" -a "$(DISTRELEASE)" != "wheezy" ]; then dh_systemd_enable --name=mysql; dh_systemd_enable --no-enable --name=mysql@; fi
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules_zenfs]: as all supported Ubuntu now works with systemd we can modify this line and remove if statement

File: build-ps/debian/rules_zenfs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+	dh_installinit --name=mysql -- defaults 19 21
+	dh_installinit --name=mysqlrouter -- defaults 19 21
+	if [ "$(DISTRELEASE)" != "precise" -a "$(DISTRELEASE)" != "trusty" -a "$(DISTRELEASE)" != "wheezy" ]; then dh_systemd_start --restart-after-upgrade; fi
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules_zenfs]: as all supported Ubuntu now works with systemd we can modify this line and remove if statement

File: build-ps/percona-server-8.0_builder.sh [hunk 8]
```diff
@@ -281,18 +284,48 @@ get_sources(){
...
+    local mode=$1
+
+    cp $CURDIR/source_tarball/percona-server-8.0.properties $WORKDIR
+    source $WORKDIR/percona-server-8.0.properties
+
+    rm -rf storage/rocksdb/rocksdb_plugins/zenfs
+    git clone https://github.com/westerndigitalcorporation/zenfs.git storage/rocksdb/rocksdb_plugins/zenfs
+    if [[ $mode == "tarball" ]]; then
+        rm build-ps/build-binary.sh
+        curl https://raw.githubusercontent.com/Sudokamikaze/percona-server/PS-7764/build-ps/build-binary.sh --output build-ps/build-binary.sh
```
Comment (EvgeniyPatlan) [inline build-ps/percona-server-8.0_builder.sh]: please do not forget to update path before merge

File: build-ps/percona-server-8.0_builder.sh [hunk 9]
```diff
@@ -281,18 +284,48 @@ get_sources(){
...
+        chmod +x build-ps/build-binary.sh
+        git clone --recursive https://github.com/percona-ysorokin/rocksdb.git -b percona_wdc $WORKDIR/TARGET/rocksdb-source
+    elif [[ $mode == "debian" ]]; then
+        mkdir builddir
+        curl https://raw.githubusercontent.com/Sudokamikaze/percona-server/PS-7764/build-ps/debian/rules_zenfs --output debian/rules_zenfs
```
Comment (EvgeniyPatlan) [inline build-ps/percona-server-8.0_builder.sh]: why not `cp rules.zenfs rules` ?

**Review by EvgeniyPatlan — CHANGES_REQUESTED** [review]

File: build-ps/percona-server-8.0_builder.sh (hunk 9 above)
Comment (Sudokamikaze) [reply build-ps/percona-server-8.0_builder.sh]: Because this file is not present in Yura's repository

File: build-ps/debian/rules_zenfs (hunk 7 above)
Comment (Sudokamikaze) [reply build-ps/debian/rules_zenfs]: Done

File: build-ps/debian/rules_zenfs (hunk 6 above)
Comment (Sudokamikaze) [reply build-ps/debian/rules_zenfs]: Done

File: build-ps/debian/rules_zenfs (hunk 5 above)
Comment (Sudokamikaze) [reply build-ps/debian/rules_zenfs]: Please compare with rules, you will see that there's no indentation problem, sent screenshot into slack

File: build-ps/debian/rules_zenfs (hunk 4 above)
Comment (Sudokamikaze) [reply build-ps/debian/rules_zenfs]: Please compare with rules, you will see that there's no indentation problem, sent screenshot into slack

File: build-ps/debian/rules_zenfs (hunk 3 above)
Comment (Sudokamikaze) [reply build-ps/debian/rules_zenfs:96]: Please compare with rules, you will see that there's no indentation problem, sent screenshot into slack

File: build-ps/debian/rules_zenfs (hunk 2 above)
Comment (Sudokamikaze) [reply build-ps/debian/rules_zenfs]: Please compare with rules, you will see that there's no indentation problem, sent screenshot into slack

File: build-ps/debian/rules [hunk 10]
```diff
@@ -201,8 +201,8 @@ override_dh_installinit:
 	@echo "RULES.$@"
 #	dh_apparmor -ppercona-server-server --profile-name=usr.sbin.mysqld
 #	dh_apparmor -ppercona-mysql-router --profile-name=usr.bin.mysqlrouter
-	if [ "$(DISTRELEASE)" != "precise" -a "$(DISTRELEASE)" != "trusty" -a "$(DISTRELEASE)" != "wheezy" ]; then dh_systemd_enable --name=mysql; dh_systemd_enable --no-enable --name=mysql@; fi
+	dh_systemd_enable --name=mysql; dh_systemd_enable --no-enable --name=mysql@; fi
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules]: fi at the end

File: build-ps/debian/rules [hunk 11]
```diff
@@ -201,8 +201,8 @@ override_dh_installinit:
...
 	dh_installinit --name=mysql -- defaults 19 21
 	dh_installinit --name=mysqlrouter -- defaults 19 21
-	if [ "$(DISTRELEASE)" != "precise" -a "$(DISTRELEASE)" != "trusty" -a "$(DISTRELEASE)" != "wheezy" ]; then dh_systemd_start --restart-after-upgrade; fi
+	dh_systemd_start --restart-after-upgrade; fi
```
Comment (EvgeniyPatlan) [inline build-ps/debian/rules]: fi at the end

**Review by EvgeniyPatlan — CHANGES_REQUESTED** [review]

**Review by EvgeniyPatlan — APPROVED** [review]

**Pull request merged**